3.6/release-notes-new-features36.md

```diff
@@ -12,12 +12,128 @@ here.
+form of:
+
+```
+<array> [ <ALL|ANY|NONE> ] [ <=|<|==|!=|>|>=|IN ] doc.<attribute>
+```
+
+i.e. the left-hand side operand is always an array, which can be dynamic.
+
+```js
+LET tokens = TOKENS("some input", "text_en")
+FOR doc IN myView SEARCH tokens ALL IN doc.title RETURN d   // dynamic conjunction
+FOR doc IN myView SEARCH tokens ANY IN doc.title RETURN d   // dynamic disjunction
+FOR doc IN myView SEARCH tokens NONE IN doc.title RETURN d  // dynamic negation
+FOR doc IN myView SEARCH tokens ALL > doc.title RETURN d    // dynamic conjunction with comparison
+FOR doc IN myView SEARCH tokens ANY <= doc.title RETURN d   // dynamic disjunction with comparison
+```
+
+In addition, both the `TOKENS()` and the `PHRASE()` functions were
+extended with array support.
+
+`TOKENS()` accepts recursive arrays of strings as the first argument:
+
+```js
+RETURN TOKENS("quick brown fox", "text_en")        // returns [ "quick", "brown", "fox" ]
+RETURN TOKENS(["quick brown", "fox"], "text_en")   // returns [ ["quick brown"], ["fox"] ]
+RETURN TOKENS(["quick brown", ["fox"]], "text_en") // returns [ ["quick brown"], [["fox"]] ]
+```
+
+`PHRASE()` accepts an array as the second argument:
+
+```js
+LET tokens = TOKENS("quick brown fox", "text_en") // ["quick", "brown", "fox"]
+FOR doc IN myView SEARCH PHRASE(doc.title, tokens, "text_en") RETURN d
+```
+
+### Analyzers
+
+Added UTF-8 support and ability to mark beginning/end of the sequence to `ngram` Analyzer
+
+The following optional properties can be provided with `ngram` analyzer definition:
+
+- `startMarker` : `<string>`, default: ""<br>
+  this value will be prepended to ngrams at the beginning of input sequence
+
+- `endMarker` : `<string>`, default: ""<br>
+  this value will be appended to ngrams at the beginning of input sequence
+
+- `streamType` : `"binary"|"utf8"`, default: "binary"<br>
+  type of the input stream
+
+### Added "Edge ngram" support to `text` analyzer
+
+The following optional properties can be provided with `text`
+Analyzer definition:
+
+- `edgeNgram` : <object>, default - not set
+  - `min` : <number>, default: not set<br>
+    min ngram length
+  - `max` : <number>, default: not set<br>
+    max ngram length
+  - `preserveOriginal`: <bool>, default: not set<br>
+    preserve original token if its length < min or > max
+
 AQL
 ---
 
+### Late document materialization
+
+With "Late document materialization" optimization ArangoDB tries to
+read only documents that are absolutely necessary for query result
+reducing load to the storage engine.
+
+In 3.6 the optimization can only be applied to the queries containing
+`SORT`+`LIMIT` combination, e.g:
+
+```js
+FOR d IN documentSource // documentSource can be either collection or ArangoSearch View
+SORT d.foo
+LIMIT 100
+RETURN d
+```
+
+For the collection case the optimization is possible IFF:
+- there is an index of type `hash`, `skip` or `edge` picked by the optimizer
```

Comment:
```suggestion
- there is an index of type `hash`, `skiplist` or `edge` picked by the optimizer
```
Does this also work with indexes of type "persistent"? At least for the RocksDB storage engine, the "persistent" index is identical to hash & skiplist indexes in terms of functionality, so it should also be supported here.

Reply:
Also what about the `primary` index?

Reply:
1. The optimization works for rocksdb only. So "persistent" is excluded.
2. Primary index is on:
`FOR d IN collectionName FILTER d._key IN ['c0', 'c1'] SORT NOOPT(d._key) DESC LIMIT 1 RETURN d`.

Reply:
RocksDB has a persistent index. Why is it excluded?

Reply:
Sorry, it is not hasCovering(). Forget about rocksdb.

Reply:
> The optimization works for rocksdb only.

No late materialization if you use MMFiles storage engine, makes sense.

> So "persistent" is excluded. / it is not hasCovering().

I'm confused about this. In RocksDB, the index types `hash`, `skiplist` and `persistent` are identical and we even dropped the first two in the Web UI. Does it work with all three index type? In MMFiles, I don't expect any index type including `persistent` to work of course.

Reply:
@Simran-B You are right. The persistent index is included also. Thank you!